gui/liquids.lua

```diff
@@ -1,336 +1,195 @@
-    elseif flow.temp_flow_timer > 0 then
-        dc:string("Flowing "..flow.temp_flow_timer, COLOR_GREEN)
-    else
-        dc:string("No Flow")
+function SpawnLiquid:spawn(pos)
+    if dfhack.maps.isValidTilePos(pos) and dfhack.maps.isTileVisible(pos) then
+        liquids.spawnLiquid(pos, self.level, self.type)
     end
 end
 
-function LiquidsUI:onRenderBody(dc)
-    dc:clear():seek(1,1):string("Paint Liquids Cheat", COLOR_WHITE)
+function SpawnLiquid:getPen()
+    local tile = nil
 
-    local cursor = guidm.getCursorPos()
-    local block = dfhack.maps.getTileBlock(cursor)
-
-    if block then
-        local x, y = pos2xyz(cursor)
-        local tile = block.tiletype[x%16][y%16]
-
-        dc:seek(2,3):string(df.tiletype.attrs[tile].caption, COLOR_CYAN)
-        dc:newline(2):pen(COLOR_DARKGREY)
-        render_liquid(dc, block, x, y)
-        dc:newline(2):pen(COLOR_DARKGREY)
-        render_flow_state(dc, block, x, y)
+    if self.type == df.tile_liquid.Water then
+        tile = dfhack.screen.findGraphicsTile('MINING_INDICATORS', 0, 0)
     else
-        dc:seek(2,3):string("No map data", COLOR_RED):advance(0,2)
+        tile = dfhack.screen.findGraphicsTile('MINING_INDICATORS', 1, 0)
     end
 
-    dc:newline():pen(COLOR_GREY)
+    return self.type == df.tile_liquid.Water and COLOR_BLUE or COLOR_RED, "X", tile
+end
 
-    dc:newline(1):key('CUSTOM_B'):string(', '):key('CUSTOM_SHIFT_B'):string(": ")
-    self.brush:render(dc)
-    dc:newline(1):key('CUSTOM_P'):string(', '):key('CUSTOM_SHIFT_P'):string(": ")
-    self.paint:render(dc)
+function SpawnLiquid:getBounds(start_position, end_position)
+    return {
+        x1=math.min(start_position.x, end_position.x),
+        x2=math.max(start_position.x, end_position.x),
+        y1=math.min(start_position.y, end_position.y),
+        y2=math.max(start_position.y, end_position.y),
+        z1=math.min(start_position.z, end_position.z),
+        z2=math.max(start_position.z, end_position.z),
+    }
+end
 
-    local paint = self.paint:get()
+function SpawnLiquid:onRenderFrame(dc, rect)
+    SpawnLiquid.super.onRenderFrame(self, dc, rect)
 
-    dc:newline()
-    if paint.liquid then
-        dc:newline(1):string("Amount: "..self.amount)
-        dc:advance(1):string("(")
-            :key('SECONDSCROLL_UP'):key('SECONDSCROLL_DOWN')
-            :string(", ")
-            :key('STRING_A048'):string("-"):key('STRING_A055')
-            :string(")")
-        dc:newline(3):key('CUSTOM_S'):string(": ")
-        self.set:render(dc)
-    else
-        dc:advance(0,2)
-    end
+    if self.is_dragging then
+        if df.global.enabler.mouse_lbut == 0 then
+            self.is_dragging = false
+        end
 
-    dc:newline()
-    if paint.flow then
-        dc:newline(1):key('CUSTOM_F'):string(": ")
-        self.flow:render(dc)
-        dc:newline(1):key('CUSTOM_R'):string(": ")
-        self.permaflow:render(dc)
-    else
-        dc:advance(0,2)
+        if not self:getMouseFramePos() then
```

Comment:
maybe this should be an `elseif` since if the button has been released, then liquids should not be spawned